capstone/capapi/views.py

```diff
@@ -56,19 +61,118 @@ class CitationViewSet(viewsets.GenericViewSet, mixins.RetrieveModelMixin, mixins
     filter_class = filters.CaseFilter
 
+    def download_many(self):
+        all_cases = self.queryset.all()
+        # get max number of cases to download from query
+        # set max number to daily max if unspecified
+        max_num = int(self.request.query_params.get('max', settings.API_CASE_DAILY_ALLOWANCE))
+
+        try:
+            to_filter = view_helpers.generate_filters_from_query_params(self.request.query_params)
+            cases = all_cases.filter(to_filter).order_by('decision_date')
+        except TypeError:
+            cases = all_cases.order_by('decision_date')
+
+        if cases.count() == 0:
+            return JsonResponse({
+                'message': 'Request did not return any results.'
+            }, status=404, )
+
+        # get page number for paginated requests
+        page_num = int(self.request.query_params.get('page', 1))
+
+        # pagination or max number
+        if max_num <= settings.API_DOWNLOAD_LIMIT:
+            cases = cases[:max_num]
+        else:
+            start_at = (page_num-1) * settings.API_DOWNLOAD_LIMIT
+            end_at = page_num * settings.API_DOWNLOAD_LIMIT
+            cases = cases[start_at:end_at]
+
+        # if request was already paginated
+        whitelisted_filters = view_helpers.get_whitelisted_case_filters()
+        blacklisted_case_count = len(set(cases) & set(all_cases.exclude(whitelisted_filters)))
+        case_allowance_sufficient = self.check_case_allowance(blacklisted_case_count)
+
+        return self.create_download_response(list(cases), blacklisted_case_count, permitted=case_allowance_sufficient)
+
+    def download_one(self, **kwargs):
+        """
+        If downloading a single case (using its lookup_field) explicitly
+        """
+        try:
+            case = models.CaseMetadata.objects.get(**kwargs)
+        except ObjectDoesNotExist as e:
+            return JsonResponse({'message': 'Unable to find case with matching slug: %s' % e}, status=404, )
+
+        blacklisted_case_count = 0 if case.jurisdiction.name in OPEN_CASE_JURISDICTIONS else 1
+        case_allowance_sufficient = self.check_case_allowance(blacklisted_case_count)
+
+        return self.create_download_response([case], blacklisted_case_count, permitted=case_allowance_sufficient)
+
+    def create_download_response(self, case_list, blacklisted_case_count, permitted=False):
+        if permitted:
+            try:
+                zip_filename = self.get_zip_filename(case_list)
+
+                # update case allowance
+                self.request.user.update_case_allowance(case_count=blacklisted_case_count)
+
+                response = StreamingHttpResponse(FileWrapper(open(zip_filename, 'rb')), content_type='application/zip')
+                response['Content-Length'] = os.path.getsize(zip_filename)
+                response['Content-Disposition'] = 'attachment; filename="%s"' % zip_filename
+                return response
+            except Exception as e:
```

Comment:
I don't think we want a try-except here -- just show a 500 if something goes wrong.